src/Data/Avro/Encoding/FromAvro.hs

```diff
@@ -0,0 +1,354 @@
+  {-# INLINE fromAvro #-}
+
+instance FromAvro Time.Day where
+  fromAvro (Int (ReadSchema.Int (Just ReadSchema.Date)) n) = Right $ fromDaysSinceEpoch (toInteger n)
+  fromAvro x                                               = Left ("Unable to decode Day from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance FromAvro Time.DiffTime where
+  fromAvro (Int (ReadSchema.Int (Just ReadSchema.TimeMillis)) n)          = Right $ millisToDiffTime (toInteger n)
+  fromAvro (Long (ReadSchema.Long _ (Just ReadSchema.TimestampMillis)) n) = Right $ millisToDiffTime (toInteger n)
+  fromAvro (Long (ReadSchema.Long _ (Just ReadSchema.TimeMicros)) n)      = Right $ microsToDiffTime (toInteger n)
+  fromAvro (Long (ReadSchema.Long _ (Just ReadSchema.TimestampMicros)) n) = Right $ microsToDiffTime (toInteger n)
+  fromAvro x                                                              = Left ("Unable to decode TimeDiff from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance FromAvro Time.UTCTime where
+  fromAvro (Long (ReadSchema.Long _ (Just ReadSchema.TimestampMicros)) n) = Right $ microsToUTCTime (toInteger n)
+  fromAvro (Long (ReadSchema.Long _ (Just ReadSchema.TimestampMillis)) n) = Right $ millisToUTCTime (toInteger n)
+  fromAvro x                                                              = Left ("Unable to decode UTCTime from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance FromAvro a => FromAvro [a] where
+  fromAvro (Array vec) = mapM fromAvro $ V.toList vec
+  fromAvro x           = Left ("Unable to decode Array from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance FromAvro a => FromAvro (Vector a) where
+  fromAvro (Array vec) = mapM fromAvro vec
+  fromAvro x           = Left ("Unable to decode Array from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance (UV.Unbox a, FromAvro a) => FromAvro (UV.Vector a) where
+  fromAvro (Array vec) = UV.convert <$> mapM fromAvro vec
+  fromAvro x           = Left ("Unable to decode Array from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance FromAvro a => FromAvro (Identity a) where
+  fromAvro (Union _ 0 v) = Identity <$> fromAvro v
+  fromAvro (Union _ n _) = Left ("Unable to decode Identity value from value with a position #" <> show n)
+  fromAvro x             = Left ("Unable to decode Identity from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance FromAvro a => FromAvro (Maybe a) where
+  fromAvro (Union _ _ Null) = Right Nothing
+  fromAvro (Union _ _ v)    = Just <$> fromAvro v
+  fromAvro x                = Left ("Unable to decode Maybe from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance (FromAvro a, FromAvro b) => FromAvro (Either a b) where
+  fromAvro (Union _ 0 a) = Left <$> fromAvro a
+  fromAvro (Union _ 1 b) = Right <$> fromAvro b
+  fromAvro (Union _ n _) = Left ("Unable to decode union value with a position #" <> show n)
```

Comment:
I wonder if this error message should be something about decoding `Either` _from_ a union value.